src/zkey/ep11.c

```diff
@@ -402,3 +446,421 @@ int reencipher_ep11_key(struct ep11_lib *ep11, target_t target,
+	if (rv != CKR_OK) {
+		DEBUG("Error m_GetAttributeValue, rv = 0x%lx", rv);
+		goto ret;
+	}
+
+	memcpy(public_key, point, point_len);
+	*publen = point_len;
+
+	rc = 0;
+
+ret:
+
+	return rc;
+}
+
+int ec_key_generate_ep11(struct ep11_lib *ep11, int curve,
+				unsigned int flags,
+				unsigned char *secure_key, unsigned int *seclen,
+				unsigned char *public_key, unsigned int *publen,
+				target_t target)
+{
+	unsigned char *ep11_pin_blob = NULL;
+	CK_ULONG ep11_pin_blob_len = 0;
+	CK_MECHANISM mech = { CKM_EC_KEY_PAIR_GEN, NULL, 0 };
+	CK_BYTE blob[MAX_BLOBSIZE];
+	size_t blob_len = MAX_BLOBSIZE;
+	unsigned char spki[MAX_BLOBSIZE];
+	size_t spki_len = sizeof(spki);
+	int rc;
+	CK_RV rv;
+	CK_BYTE *q;
+	size_t qlen;
+	CK_BYTE p256_params[] = ZPC_P256_PARAMS;
+	CK_BYTE p384_params[] = ZPC_P384_PARAMS;
+	CK_BYTE p521_params[] = ZPC_P521_PARAMS;
+	CK_BYTE ed25519_params[] = ZPC_ED25519_PARAMS;
+	CK_BYTE ed448_params[] = ZPC_ED448_PARAMS;
+	CK_ULONG ec_params_len_from_curve[] = {
+		sizeof(p256_params), sizeof(p384_params), sizeof(p521_params),
+		sizeof(ed25519_params), sizeof(ed448_params),
+	};
+	CK_BYTE *ec_params;
+	CK_ULONG ec_params_len = 0;
+	int MAX_EP11_ATTRS = 15;
+	CK_ATTRIBUTE priv_attrs[MAX_EP11_ATTRS];
+	CK_BBOOL cktrue = CK_TRUE;
+
+
+	if (ep11->dll_m_GenerateKeyPair == NULL) {
+		DEBUG("EP11 host lib function m_GenerateKeyPair not "
+			"available but required for EP11 EC key generate.");
+		return -EIO;
+	}
+
+	switch (curve) {
+	case 0:
+		ec_params = &p256_params[0];
+		break;
+	case 1:
+		ec_params = &p384_params[0];
+		break;
+	case 2:
+		ec_params = &p521_params[0];
+		break;
+	case 3:
+		ec_params = &ed25519_params[0];
+		break;
+	case 4:
+		ec_params = &ed448_params[0];
+		break;
+	}
+	ec_params_len = ec_params_len_from_curve[curve];
+
+	CK_ATTRIBUTE default_priv_attrs[] = {
+		{CKA_IBM_PROTKEY_EXTRACTABLE, &cktrue, sizeof(cktrue)},
+		{CKA_PRIVATE, &cktrue, sizeof(cktrue)},
+		{CKA_SIGN, &cktrue, sizeof(cktrue)},
+	};
+	CK_ATTRIBUTE pub_attrs[] = {
+		{ CKA_VERIFY, &cktrue, sizeof(cktrue) },
+		{ CKA_EC_PARAMS, ec_params, ec_params_len },
+	};
+	CK_ULONG pub_attrs_len = sizeof(pub_attrs) / sizeof(CK_ATTRIBUTE);
+	CK_ULONG priv_attrs_len;
+	struct ep11kblob_header *ep11hdr;
+
+	/* Add default attributes */
+	memcpy(&priv_attrs, &default_priv_attrs, sizeof(default_priv_attrs));
+	priv_attrs_len = sizeof(default_priv_attrs) / sizeof(CK_ATTRIBUTE);
+
+	/* Add user-defined flags, translated to attributes */
+	ec_key_add_attrs_from_flags(flags, priv_attrs, &priv_attrs_len, &cktrue);
+
+	/* Generate the secure key blob */
+	rv = ep11->dll_m_GenerateKeyPair(&mech, pub_attrs, pub_attrs_len,
+						priv_attrs, priv_attrs_len,
+						ep11_pin_blob, ep11_pin_blob_len,
+						blob, &blob_len,
+						spki, &spki_len, target);
+	if (rv != CKR_OK) {
+		DEBUG("Error m_GenerateKeyPair, rv = 0x%lx", rv);
+		rc = EIO;
+		goto ret;
+	}
+
+	/* Add info to ep11 header (overlay over session ID field). This info
```

Comment:
here you still mention 'overlay' ? I guess thats no longer true, right?

Reply:
hm, looks like you are still overlaying here, so this needs to be changed to a blob with a prepended header!

Reply:
Code is reworked. No overlay anymore except of one place: in aes_key.c, routine aes_key_sec2prot_with_header the ioctl wants a TOKEN_VERSION_EP11_AES (i.e. no header). But there I'm saving the original data in the first 16 bytes and I'm restoring them after the ioctl.

Reply:
> ioctl wants a TOKEN_VERSION_EP11_AES

Isn't there a newer ioctl command available that works on tokens with external header?

Reply:
Unfortunately not. But I noted the point for the future.